README.md
![i-sense flu](https://res.cloudinary.com/uclfmedia/image/upload/v1563449524/isenseflu/logo_300.svg)

# i-sense flu publish

Module of the i-sense flu application used in the scheduled publication of model scores on social media (Twitter)

[![Build Status](https://travis-ci.org/UCL/isenseflu-publish-javaee.svg?branch=master)](https://travis-ci.org/UCL/isenseflu-publish-javaee)
![GitHub release](https://img.shields.io/github/release/UCL/isenseflu-publish-javaee.svg)


## Background

i-sense flu is a multi-module application that uses Google search data to 
estimate influenza-like illness (flu) rates in England. The i-sense flu publish 
module is responsible for publishing a tweet containing the latest score, 
along with a chart showing the scores for the last 30 days, and the rate of 
change in the scores of the last 28 days, relative to the previous 28-day period.


## Features

- Scheduled publishing of tweets based on a set template
- Creation of a chart containing the scores for the last 30 days
- Scores are received via STOMP messages
- Allows the reception of one score per day


## Technologies

- Java EE8
- STOMP messaging
- REST


## Installation

### Requirements

[Eclipse Glassfish 5][eclipse-glassfish-5] full profile (version 5.1.0) running OpenJDK 8

If using [Glassfish version 5.0][javaee-glassfish-5], make sure that the JDK/JRE version is 1.8.0 u151

```
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)
```

Ref: [eclipse-ee4j/glassfish#22436](https://github.com/eclipse-ee4j/glassfish/issues/22436)

> There is some incompatible changes in the JDK after u152 that is impacting 
Grizzly. Hence, Grizzly had to be fixed and integrated into GlassFish 5.0.1.

### Configuration

Before deploying the application for the first time, the following steps are required to integrate the application with the i-sense flu API and Twitter API.

#### Aliases for credentials

Passwords and API keys/tokens need to be configured in Glassfish via aliases instead of clear text. The following aliases need to be created:

Consumer key and access token credentials from [Twitter app dashboard][twitter-app-dashboard]:

- `twitter-apikey-alias`
- `twitter-apisecret-alias`
- `twitter-token-alias`
- `twitter-tokensecret-alias`

And the admin user credential in the message queue (OpenMQ) to be used by the STOMP bridge service:

- `imq-adminpass-alias`

Configure each one with the Glassfish CLI admin tool `asadmin`. For example:

```
asadmin> create-password-alias twitter-apikey-alias
```

#### Message Queue (OpenMQ)

The i-sense flu API component is scheduled to send the latest score via a STOMP message. Glassfish provides an installation of OpenMQ that supports STOMP messaging via a JMS bridge. This can be enabled and configured with `asadmin`

```
asadmin> set configs.config.server-config.jms-service.type=LOCAL
asadmin> set configs.config.server-config.jms-service.property.imq\\.jmsra\\.managed=true
asadmin> set configs.config.server-config.jms-service.property.imq\\.bridge\\.enabled=true
asadmin> set configs.config.server-config.jms-service.property.imq\\.bridge\\.admin\\.user=[username]
asadmin> set configs.config.server-config.jms-service.property.imq\\.bridge\\.admin\\.password=${ALIAS\\=imq-adminpass-alias}
asadmin> set configs.config.server-config.jms-service.property.imq\\.bridge\\.activelist=stomp
```

Restart the Glassfish domain to activate the configuration

```
asadmin> restart-domain
```

#### System properties/environment variables

The following system properties are required to be configured in the application server:

- `API_SCORES_URI`: URI of the `/scores` REST endpoint of the i-sense flu API component
- `TWITTER_SCHEDULED_FOR`: Time when the Tweet with the latest score report should be published. In `hh:mm` format.

 

- `TWITTER_KEY`: Consumer API key for Twitter set as `twitter-apikey-alias` password alias
- `TWITTER_SECRET`: Consumer API secret key for Twitter set as `twitter-apisecret-alias` password alias
- `TWITTER_TOKEN`: Access token for Twitter set as `twitter-token-alias` password alias
- `TWITTER_TOKEN_SECRET`: Access token secret for Twitter set as `twitter-tokensecret-alias` password alias

In Glassfish these system properties can be configured as follows:

```
asadmin> create-system-properties API_SCORES_URI=http\\://localhost\\:8080/api/flu/scores
asadmin> create-system-properties TWITTER_SCHEDULED_FOR=16\\:00
asadmin> create-system-properties TWITTER_KEY='${ALIAS\=twitter-apikey-alias}'
```

Alternatively, these values can be configured via environment variables in the shell. This way of configuration can be used when deploying to environments like Openshift or Heroku.

### Deployment

Download the latest release (or build) and deploy the EAR application using the `asadmin` utility from Glassfish:

```
asadmin> deploy --name isenseflu-publish [...path/to/...]isenseflu-ear-[version].ear
```

Or via Administration Console (usually `http://hostname:4848`)


## Building

Build the application from source using the `mvnw` Maven wrapper script provided, it requires `JAVA_HOME` set to JDK8 (tested with Open JDK 8). Fetch the sources from the master branch only.

```
./mvnw package
```

The EAR application will be found under `isenseflu-publish-ear/target/`

The system test runs within a local installation of Glassfish 5.1 and uses Mock Server to mock the external APIs. The tests follow the process from receiving the STOMP message up to publishing the Tweet in the required format. The complete system test should not take more than 3 minutes to complete. Run with:

```
./mvnw verify
```

## Reporting bugs

Please use the GitHub issue tracker for any bugs or feature suggestions:

[https://github.com/UCL/isenseflu-publish-javaee/issues](https://github.com/UCL/isenseflu-publish-javaee/issues)


## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate. Analyse your code with [CheckStyle][checkstyle] (Sun Checks) and [Findbugs/Spotbugs][findbugs].

Please keep the use of 3rd party runtime libraries to a minimum. The application is expected to run on a standard Java EE 8 server ([Eclipse Glassfish 5 full profle][eclipse-glassfish-5]). 


## Authors

- David Guzman (Github: [@david-guzman](https://github.com/david-guzman))


## Acknowledgements

i-sense flu publish is supported by the EPSRC IRC project [i-sense](https://www.i-sense.org.uk/) (Early-Warning Sensing Systems for Infectious Diseases).


## Copyright

isenseflu-publish-javaee is licensed under the GNU General Public License, v3. A copy of this license is included in the file [LICENSE.md](LICENSE.md).


&copy; 2019 UCL ([https://www.ucl.ac.uk](https://www.ucl.ac.uk)).


[eclipse-glassfish-5]: https://projects.eclipse.org/projects/ee4j.glassfish/downloads
[checkstyle]: https://checkstyle.org/
[findbugs]: https://github.com/findbugsproject/findbugs
[javaee-glassfish-5]: https://javaee.github.io/glassfish/
[twitter-app-dashboard]: https://developer.twitter.com/en/apps


CHANGELOG.md
# Changelog

## Latest release: [1.1.0](https://github.com/UCL/isenseflu-publish-javaee/releases/tag/v1.1.0)

### Changed

- Average increase/decrease rate now calculated based on the scores from the last 7 days compared to the scores from the preceding 28-day period.


isenseflu-publish-ejb/src/main/java/uk/ac/ucl/isenseflu/publish/MessageParser.java
/*
 * i-sense flu publish: Module of the i-sense flu application used in the
 * publication of model scores on social media
 *
 * Copyright (c) 2019, UCL <https://www.ucl.ac.uk/>
 *
 * This file is part of i-sense flu publish
 *
 * i-sense flu publish is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * i-sense flu publish is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with i-sense flu publish.  If not, see <http://www.gnu.org/licenses/>.
 */

package uk.ac.ucl.isenseflu.publish;

import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.MessageFormat;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Properties;
import java.util.UUID;
import java.util.stream.Collectors;
import javax.enterprise.context.Dependent;
import javax.imageio.ImageIO;
import javax.inject.Inject;

/**
 * CDI bean responsible of the generation of the content of the tweet
 * from the message containing the date and the value of the score.
 * Since Twitter offers a two-step process to include images in a tweei,
 * MessageParser uses the TweetData as a container for both the
 * message text and the image.
 *
 * @author David Guzman
 */
@Dependent
public class MessageParser {

  /**
   * Ordinal suffix to use with days ending in 1.
   */
  private static final String FIRST_SUFFIX = "st of ";

  /**
   * Ordinal suffix to use with days ending in 2.
   */
  private static final String SECOND_SUFFIX = "nd of ";

  /**
   * Ordinal suffix to use with days ending in 3.
   */
  private static final String THIRD_SUFFIX = "rd of ";

  /**
   * Number of decimal places that the score should use in a tweet.
   */
  private static final int SCORE_DECIMAL_PLACES = 3;

  /**
   * For calculation of percentages.
   */
  private static final float PERCENT = 0.01f;

  /**
   * Divisor used in the extraction of the last digit of a day number.
   */
  private static final int LAST_DIGIT_IN_DAY_DIVISOR = 10;

  /**
   * Remainder for days ending in 1.
   */
  private static final int FIRST_OF_MODULUS = 1;

  /**
   * Remainder for days ending in 2.
   */
  private static final int SECOND_OF_MODULUS = 2;

  /**
   * Remainder for days ending in 3.
   */
  private static final int THIRD_OF_MODULUS = 3;

  /**
   * Bean responsible for the retrieval of flu scores from the i-sense flu API.
   */
  @Inject
  private FetchScores fluDetectorScores;

  /**
   * Bean responsible of the generation of the chart to include in the tweet.
   */
  @Inject
  private PlotModelScore plotModelScore;

  /**
   * Template of the tweet. It requires the following parameters, in order:
   * the date of the score, the value of the score, whether it's an increase
   * or decrease, the relative change rate, the start date of the chart, and
   * the end date of the chart.
   */
  private final MessageFormat tweetFormat = new MessageFormat(
    "Based on Google searches, the estimated flu (influenza-like illness) rate"
      + " for England on the {0} was {1} cases per 100,000 people with an "
      + "average 7-day {2} rate of {3}% compared to the previous 28-day "
      + "period https://www.i-senseflu.org.uk/?start={4}&end={5}&resolution=day"
      + "&smoothing=0&id=3&source=twlink #health #AI");

  /**
   * Generates the text of the tweet and the chart image to publish on Twitter.
   * It uses a one month period before the date of the score to provide the
   * data for the chart and rate change, calling FetchScores for the retrieval
   * of the scores for the previous 56 days (sorted by date in descending order)
   * and PlotModelScore to generate the chart.
   * @param message A string containing the score data in a format that can
   *                be read as Properties. The properties required are: date
   *                (in ISO format) and value (a number).
   * @return        An instance of TweetData, containing the text of the tweet
   *                and the chart image.
   */
  public TweetData getTweetData(final String message) {
    Properties properties = PropertyReader.readProperties(message)
            .orElseThrow(
              () -> new IllegalArgumentException(
                "Cannot read properties from message"
              )
            );

    String date = properties.getProperty("date", "");
    String value = properties.getProperty("value", "");

    if (date.isEmpty() || !date.matches("^\\d{4}-\\d{2}-\\d{2}$")) {
      throw new IllegalArgumentException(
        "Date field not provided or date not in ISO format: date is "
          + date
      );
    }

    if (value.isEmpty() || !value.matches("\\d+(\\.\\d+)?")) {
      throw new IllegalArgumentException(
        "Score value not provided or score value not a number: value is "
          + value
      );
    }

    LocalDate endDate = LocalDate.parse(date);
    LocalDate startDate = LocalDate.parse(date).minusMonths(1);
    String ordinal = "th of ";
    int day = endDate.getDayOfMonth();
    switch (day % LAST_DIGIT_IN_DAY_DIVISOR) {
      case FIRST_OF_MODULUS:
        ordinal = FIRST_SUFFIX;
        break;
      case SECOND_OF_MODULUS:
        ordinal = SECOND_SUFFIX;
        break;
      case THIRD_OF_MODULUS:
        ordinal = THIRD_SUFFIX;
        break;
      default:
        break;
    }
    String formattedDate = endDate
            .format(DateTimeFormatter.ofLocalizedDate(
              FormatStyle.LONG).withLocale(Locale.UK)
            )
            .replaceFirst("^0", "")
            .replaceAll("-", " ")
            .replaceFirst(" ", ordinal)
            .replaceFirst("( \\d{4})$", ",$1");

    BigDecimal bd = new BigDecimal(value);
    bd = bd.setScale(SCORE_DECIMAL_PLACES, RoundingMode.HALF_UP);

    List<DatapointModelScore> scoresList = fluDetectorScores
      .getScoresForLast56Days(endDate);

    // Use a copy last 30 days from scoresList. Not thread-safe
    BufferedImage chart = plotModelScore.createLineChart(
      new ArrayList<>(scoresList.subList(0, 30))
    );

    List<Double> scores = scoresList.stream()
            .map((DatapointModelScore s) -> s.getScoreValue())
            .collect(Collectors.toList());
    Double p = RateCalculator.averageChangeRate().apply(scores);
    String variation = "variation";
    if (p > 0) {
      variation = "increase";
    } else if (p < 0) {
      variation = "decrease";
    }

    Object[] toFormat = {
      formattedDate,
      bd.toPlainString(),
      variation,
      Math.abs(p / PERCENT),
      startDate,
      endDate
    };
    String tweet = tweetFormat.format(toFormat);

    return new TweetData(tweet, chart);
  }

  public static final class TweetData {

    /**
     * The generated text of the tweet.
     */
    private final String tweet;

    /**
     * The chart showing the scores for the last month of data.
     */
    private final BufferedImage chart;

    /**
     * Identifier used to name the resulting chart image.
     */
    private final UUID chartId = UUID.randomUUID();

    private TweetData(final String tweetStr, final BufferedImage chartImg) {
      this.tweet = tweetStr;
      this.chart = chartImg;
    }

    /**
     * Accesses the text of the tweet.
     * @return The text content of the tweet.
     */
    public String getTweet() {
      return tweet;
    }

    /**
     * Produces a png image of the chart.
     * @return              The chart image ss an InputStream.
     * @throws IOException  If the conversion from a BufferedImage fails.
     */
    public InputStream getChartAsPng() throws IOException {
      final ByteArrayOutputStream os = new ByteArrayOutputStream() {
        @Override
        public synchronized byte[] toByteArray() {
          return this.buf;
        }
      };
      ImageIO.write(chart, "png", os);
      return new ByteArrayInputStream(os.toByteArray(), 0, os.size());
    }

    /**
     * Appends the png extension to use with the chart image file.
     * @return The name of the chart image file.
     */
    public String getPngFilename() {
      return chartId + ".png";
    }

  }

}
